master/internal/token/postgres_token.go

```diff
@@ -0,0 +1,154 @@
+// user_sessions db.
+func CreateAccessToken(
+	ctx context.Context, userID model.UserID, opts ...AccessTokenOption,
+) (string, model.TokenID, error) {
+	CurrentTimeNowInUTC = time.Now().UTC()
+	// Populate the default values in the model.
+	accessToken := &model.UserSession{
+		UserID:      userID,
+		CreatedAt:   CurrentTimeNowInUTC,
+		Expiry:      CurrentTimeNowInUTC.Add(DefaultTokenLifespan),
+		TokenType:   model.TokenTypeAccessToken,
+		Description: null.StringFromPtr(nil),
+		Revoked:     false,
+	}
+
+	// Update the optional ExpiresAt field (if passed)
+	for _, opt := range opts {
+		opt(accessToken)
+	}
+
+	var token string
+
+	err := db.Bun().RunInTx(ctx, nil, func(ctx context.Context, tx bun.Tx) error {
+		// AccessTokens should have a many:1 relationship with users.
+		// A new row is inserted into the user_sessions table, and the ID of the
+		// inserted row is returned and stored in user_sessions.ID.
+		_, err := tx.NewInsert().
+			Model(accessToken).
+			Column("user_id", "expiry", "created_at", "token_type", "revoked", "description").
+			Returning("id").
+			Exec(ctx, &accessToken.ID)
+		if err != nil {
+			return err
+		}
+
+		// A Paseto token is generated using the accessToken object and the private key.
+		v2 := paseto.NewV2()
+		privateKey := db.GetTokenKeys().PrivateKey
+		token, err = v2.Sign(privateKey, accessToken, nil)
+		if err != nil {
+			return fmt.Errorf("failed to generate user authentication token: %s", err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return "", 0, err
+	}
+
+	return token, model.TokenID(accessToken.ID), nil
+}
+
+// AccessTokenUpdateOptions is the set of mutable fields for an Access Token record.
+type AccessTokenUpdateOptions struct {
+	Description *string
+	SetRevoked  bool
+}
+
+// UpdateAccessToken updates the description and revocation status of the access token.
+func UpdateAccessToken(
+	ctx context.Context, tokenID model.TokenID, options AccessTokenUpdateOptions,
+) (*model.UserSession, error) {
+	var tokenInfo model.UserSession
+	// Lock the token row for update.
+	err := db.Bun().NewSelect().For("UPDATE").Table("user_sessions").
+		Where("id = ?", tokenID).Where("token_type = ?", model.TokenTypeAccessToken).
+		Scan(ctx, &tokenInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	if tokenInfo.Revoked || tokenInfo.Expiry.Before(time.Now().UTC()) {
+		return nil, fmt.Errorf("unable to update inactive token with ID %v", tokenID)
+	}
+
+	if options.Description != nil {
+		tokenInfo.Description = null.StringFrom(*options.Description)
+	}
+
+	if options.SetRevoked {
+		tokenInfo.Revoked = true
+	}
+
+	_, err = db.Bun().NewUpdate().
+		Model(&tokenInfo).
+		Column("description", "revoked").
+		Where("id = ?", tokenID).
+		Exec(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &tokenInfo, nil
+}
+
+// GetUserIDFromTokenID retrieves the userID associated with the provided tokenID.
+func GetUserIDFromTokenID(ctx context.Context, tokenID int32) (model.UserID, error) {
```

Comment:
nit: feel like we should use pointers here on model.UserID if it's nullable, instead of returning 0. not a big deal though.